hollow/src/main/java/com/netflix/hollow/api/consumer/metrics/AbstractRefreshMetricsListener.java

```diff
@@ -0,0 +1,178 @@
+        refreshLoadType = ConsumerRefreshLoadType.Init;
+        refreshTransitionType = null;
+    }
+
+    @Override
+    public final void refreshStarted(long currentVersion, long requestedVersion) {
+        refreshStart = Instant.now();
+        refreshTransitionType = null;   // Don't assume snapshot or delta transition, since a failed transition might
+                                        // leave the value of this variable in an incorrect state. Hence the value is
+                                        // reset at the start of every refresh.
+
+        // TODO: If this is an init load then we can get visibility into bootstrap time
+        // if (loadType.equals(ConsumerLoadType.Init)) {
+        //
+        // }
+    }
+
+    @Override
+    public final void refreshSuccessful(long beforeVersion, long afterVersion, long requestedVersion) {
+        Instant refreshEnd = Instant.now();
+
+        long refreshDurationMillis = Duration.between(refreshStart, refreshEnd).toMillis();
+        long timeSinceLastSuccessfulRefreshMillis = 0l;
+
+        consecutiveSuccesses ++;
+        consecutiveFailures = 0l;
+        // reset the timer to zero to indicate a refresh happened.
+        lastRefreshInstant = refreshEnd;
+        refreshStatus = ConsumerRefreshStatus.Success;
+
+        reportRefreshStats(new ConsumerRefreshMetrics(refreshDurationMillis, refreshStatus, refreshTransitionType, refreshLoadType));
+        reportTimeSinceLastSuccessfulRefresh(timeSinceLastSuccessfulRefreshMillis);
+        reportConsecutiveSuccesses(consecutiveSuccesses);
+        reportConsecutiveFailures(consecutiveFailures);
+    }
+
+    @Override
+    public final void refreshFailed(long beforeVersion, long afterVersion, long requestedVersion, Throwable failureCause) {
+        Instant refreshEnd = Instant.now();
+
+        long refreshDurationMillis = Duration.between(refreshStart, refreshEnd).toMillis();
+        long timeSinceLastSuccessfulRefreshMillis = Duration.between(lastRefreshInstant, refreshEnd).toMillis();
+
+        consecutiveSuccesses = 0l;
+        consecutiveFailures ++;
+        refreshStatus = ConsumerRefreshStatus.Failure;
+
+        reportRefreshStats(new ConsumerRefreshMetrics(refreshDurationMillis, refreshStatus, refreshTransitionType, refreshLoadType));
+        reportTimeSinceLastSuccessfulRefresh(timeSinceLastSuccessfulRefreshMillis);
+        reportConsecutiveSuccesses(consecutiveSuccesses);
+        reportConsecutiveFailures(consecutiveFailures);
+    }
+
+    @Override
+    public final void blobLoaded(HollowConsumer.Blob transition) {
+        refreshLoadCount ++;
+
+        if (refreshLoadCount > 1) {
+            refreshLoadType = ConsumerRefreshLoadType.Subsequent;
+        }
+    }
+
+    @Override
+    public final void snapshotUpdateOccurred(HollowAPI refreshAPI, HollowReadStateEngine stateEngine, long version) {
+        refreshTransitionType = ConsumerRefreshTransitionType.Snapshot;
```

Comment:
For the case where a consumer, in a single refresh, first applies a snapshot blob then applies a delta blob, we'll end up reporting that as a delta refresh/update once line 158 has been called.

If we're using a sampling metrics reporter (like the Specator library provides) then depending on when we sample these metrics (and assuming we're including transition type as a tag on a metric) then we might see one sample where the tag is snapshot then the subsequent sample the tag would change to delta. For example, if we report a timer metric for the entire load duration and tag that timer with transition type, then partway through that timer the transition might switch from snapshot to delta.

Reply:
This probably illustrates a fundamental gap in our current refresh listener API. There's no way for the listener to know at the onset what the update plan has in store for it, even though Hollow determines the update plan early on, long before attempting to download and apply each transition.

Reply:
Hmm.. I didn't think deltaUpdateOccurred() would be called on a "snapshot + N deltas" update plan because of https://stash.corp.netflix.com/projects/GM/repos/hollow/browse/hollow/src/main/java/com/netflix/hollow/api/client/HollowDataHolder.java#170 

That's not true for deltaApplied() though, which is the one we're watching in KeystoneConsumerListener. I do agree about not knowing snapshot-only vs. snapshot+delta(s) vs. delta(s) at the onset of refreshStart (and therefore at refreshFailure) and that might be something to look into adding.

I think we need to talk this one out as I might be missing more here than I thought. Look fwd to our in person chat.

Reply:
I'd also like to understand a bit more about the need for sampling here.

Reply:
ConsumerRefreshMetrics now has overallRefreshType set on the duration metric (it's snapshot if there is a snapshot transition in the refresh's update plan). The duration metric will continue to be reported as a Timer as it is now (as opposed to Long Task Timer or Gauge). Details of the refresh are exposed in ConsumerRefreshMetrics's UpdatePlanDetails for logging. This will be more explicit in the Cinder-side PR.